fasm/parser/antlr.py

```
@@ -0,0 +1,289 @@
+    """ Parse FASM string, returning list of FasmLine named tuples."""
+    lines = []
+    line, p = fasm_line_from_bytes(data, 0)
+    while line:
+        lines.append(line)
+        line, p = fasm_line_from_bytes(data, p)
+
+    # Check that data read, plus the final null header,
+    # is equal to the buffer size.
+    assert p + 1 == size, p
+
+    return lines
+
+
+def parse_fasm_string(s):
+    """ Parse FASM string, returning list of FasmLine named tuples.
+
+    >>> parse_fasm_string('a.b.c = 1')[0].set_feature.feature
+    'a.b.c'
+
+    Args:
+        s: The string containing FASM source to parse.
+
+    Returns:
+        A list of fasm.model.FasmLine.
+    """
+    result = [None]
+
+    # Use a closure to parse while allowing C++ to handle memory.
+    @CFUNCTYPE(None, POINTER(c_char), c_size_t)
+    def callback(s, n):
+        result[0] = parse_fasm_data(s, n)
+
+    @CFUNCTYPE(None, c_size_t, c_size_t, c_char_p)
+    def error_callback(line, position, message):
+        result[0] = None
+        print(
+            'Parse error at {}:{} - {}'.format(
+                line, position, message.decode('ascii')),
+            file=stderr)
+
+    parse_fasm.from_string(bytes(s, 'ascii'), 0, callback, error_callback)
+    return result[0]
+
+
+def parse_fasm_filename(filename):
+    """ Parse FASM file, returning list of FasmLine named tuples.
+
+    >>> parse_fasm_filename('examples/feature_only.fasm')[0]\
+        .set_feature.feature
+    'EXAMPLE_FEATURE.X0.Y0.BLAH'
+
+    Args:
+        filename: The file containing FASM source to parse.
+
+    Returns:
+        A list of fasm.model.FasmLine.
+    """
+    result = [None]
+
+    # Use a closure to parse while allowing C++ to handle memory.
+    @CFUNCTYPE(None, POINTER(c_char), c_size_t)
+    def callback(s, n):
+        result[0] = parse_fasm_data(s, n)
+
+    @CFUNCTYPE(None, c_size_t, c_size_t, c_char_p)
+    def error_callback(line, position, message):
+        result[0] = None
+        print(
```

This should throw an exception rather than print to stderr.  The exception should be throw after the `parse_fasm.from_file` (or `parse_fasm.from_str`) returns.

---

Done.